### PR TITLE
 client library integration

### PR DESCRIPTION
**Pull request by mihajlon97**
- client library
 - fiskalyServiceUrl optional parameter
 - tests adapted
 - readme updated

### CODE REVIEW

File: src/client/ClientLibrary.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    private readonly HTTP_TIMEOUT_ERROR: number = -21000;
+
+    private library: ffi.Library = null;
+
+    /**
+     * Client Library Constructor.
+     */
+    constructor() {
+        try {
+            this.library = ffi.Library(`${__dirname}\\${this.getLibraryName()}`, {
```
Comment (pgaubatz) [inline src/client/ClientLibrary.ts]: this will most likely NOT work on non-windows platforms

Comment (mihajlon97) [reply src/client/ClientLibrary.ts]: Is the library path a problem?
Should we use require('path').join()?

Comment (pgaubatz) [reply src/client/ClientLibrary.ts]: yes, please!

File: src/FiskalyClient.ts [hunk 2]
```diff
@@ -2,23 +2,33 @@ import * as jayson from 'jayson/promise';
...
-            throw new FiskalyError("fiskalyServiceUrl must be provided");
+            try {
+                this.clientLibrary = new ClientLibrary();
+                this.hasService = false;
+            } catch (e) {
+                throw new FiskalyError("fiskalyServiceUrl must be provided");
+            }
+        } else {
+            // @ts-ignore
+            this.jsonRPC = jayson.Client.http(fiskalyServiceUrl);
```
Comment (pgaubatz) [inline src/FiskalyClient.ts]: how about:
```suggestion
            try {
                const lib = new ClientLibrary();
                this.doRequest = lib.request.bind(lib);
            } catch (e) {
                throw new FiskalyError("fiskalyServiceUrl must be provided");
            }
        } else {
            const rpc = jayson.Client.http(fiskalyServiceUrl);
            this.doRequest = rpc.request.bind(rpc);
```

these class fields can also be removed then:
```ts
    private jsonRPC: jayson.Client | undefined;
    private clientLibrary: ClientLibrary | undefined;
    private hasService: boolean = true;
```

instead there would only be:
```ts
    private doRequest: Function;
```

**Review by pgaubatz — CHANGES_REQUESTED** [review]

File: src/FiskalyClient.ts [hunk 3]
```diff
@@ -2,41 +2,38 @@ import * as jayson from 'jayson/promise';
...
 import { FiskalyErrorHandler, FiskalyError } from "./errors";
 import { ConfigureMethodParams, RequestMethodParams } from "./interfaces";
+import { ClientLibrary } from "./client/ClientLibrary";
+import {JSONRPCResultLike} from "jayson/promise";
 
 export class FiskalyClient {
     private readonly SDK_VERSION = '1.1.600';
     private context: string | undefined;
-    private readonly jsonRPC: jayson.Client;
+    private readonly doRequest: OmitThisParameter<(method: string, params: object) => Promise<JSONRPCResultLike> | null> = () => null;
```
Comment (pgaubatz) [inline src/FiskalyClient.ts]:
```suggestion
    private readonly doRequestFn: OmitThisParameter<(method: string, params: object) => Promise<JSONRPCResultLike> | null> = () => null;
```

File: src/FiskalyClient.ts [hunk 4]
```diff
@@ -2,41 +2,38 @@ import * as jayson from 'jayson/promise';
...
+    private jsonRPC: jayson.Client | undefined;
+    private clientLibrary: ClientLibrary | undefined;
+    private hasService: boolean = true;
```
Comment (pgaubatz) [inline src/FiskalyClient.ts]:
```suggestion
```

File: src/FiskalyClient.ts [hunk 5]
```diff
@@ -2,41 +2,38 @@ import * as jayson from 'jayson/promise';
...
-     * @param {string} fiskalyServiceUrl
+     * If fiskalyServiceUrl parameter not provided, the client library will be loaded from 'client' folder
+     * @param {string} fiskalyServiceUrl | optional
      */
     constructor(fiskalyServiceUrl: string) {
         if (!fiskalyServiceUrl) {
-            throw new FiskalyError("fiskalyServiceUrl must be provided");
+            try {
+                const lib = new ClientLibrary();
+                this.doRequest = lib.request.bind(lib);
```
Comment (pgaubatz) [inline src/FiskalyClient.ts]:
```suggestion
                this.doRequestFn = lib.request.bind(lib);
```

File: src/FiskalyClient.ts [hunk 6]
```diff
@@ -2,41 +2,38 @@ import * as jayson from 'jayson/promise';
...
+            } catch (e) {
+                throw new FiskalyError("fiskalyServiceUrl must be provided");
+            }
+        } else {
+            // @ts-ignore
+            const rpc = jayson.Client.http(fiskalyServiceUrl);
+            this.doRequest = rpc.request.bind(rpc);
```
Comment (pgaubatz) [inline src/FiskalyClient.ts]:
```suggestion
            this.doRequestFn = rpc.request.bind(rpc);
```

File: src/FiskalyClient.ts [hunk 7]
```diff
@@ -62,8 +59,11 @@ export class FiskalyClient {
             'sdk_version': this.SDK_VERSION
         };
         const response = await this.doRequest('create-context', contextParams);
-
-        /** Update context */
+        /** Check if error exists */
+        if (response.error != null) {
+            throw FiskalyErrorHandler.throwError(response);
+        }
+        // @ts-ignore
```
Comment (pgaubatz) [inline src/FiskalyClient.ts]:
```suggestion
```

File: src/FiskalyClient.ts [hunk 8]
```diff
@@ -89,7 +89,10 @@ export class FiskalyClient {
      */
     public async getVersion(): Promise<VersionResponse> {
         const response = await this.doRequest('version', {});
-
+        /** Check if error exists */
+        if (response.error != null) {
+            throw FiskalyErrorHandler.throwError(response);
+        }
```
Comment (pgaubatz) [inline src/FiskalyClient.ts]:
```suggestion
```

File: src/FiskalyClient.ts [hunk 9]
```diff
@@ -105,7 +108,10 @@ export class FiskalyClient {
             context: this.context
         };
         const response = await this.doRequest('config', params);
-
+        /** Check if error exists */
+        if (response.error != null) {
+            throw FiskalyErrorHandler.throwError(response);
+        }
```
Comment (pgaubatz) [inline src/FiskalyClient.ts]:
```suggestion
```

File: src/FiskalyClient.ts [hunk 10]
```diff
@@ -120,6 +126,10 @@ export class FiskalyClient {
             context: this.context
         };
         const response = await this.doRequest('config', params);
+        /** Check if error exists */
+        if (response.error != null) {
+            throw FiskalyErrorHandler.throwError(response);
+        }
```
Comment (pgaubatz) [inline src/FiskalyClient.ts]:
```suggestion
```

File: src/FiskalyClient.ts [hunk 11]
```diff
@@ -138,7 +148,9 @@ export class FiskalyClient {
             context: this.context
         };
         const response = await this.doRequest('request', params);
-
+        if (response.error != null) {
+            throw FiskalyErrorHandler.throwError(response);
+        }
```
Comment (pgaubatz) [inline src/FiskalyClient.ts]:
```suggestion
```

**Review by pgaubatz — CHANGES_REQUESTED** [review]
Re-add the following method:
```ts
    /**	
     * Send JSON RPC Request to Client	
     * @param {string} method	
     * @param {object} params	
     */	
    private async doRequest(method: string, params: object) {	
        if(!this.allowedMethods.includes(method)) {	
            throw new FiskalyError("Invalid method parameter");	
        }	
        const response = await this.doRequestFn(method, params);	
        /** Check if error exists */	
        if (response.error != null) {	
            throw FiskalyErrorHandler.throwError(response);	
        }	
        return response	
    }
```

File: tests/FiskalyClientTestClientLib.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
+import { FiskalyClient } from '../src/';
+import { ClientConfiguration, VersionResponse, RequestResponse } from "../src/responses/";
+import { FiskalyHttpError } from "../src/errors";
+
+let FISKALY_SERVICE_URL: string, FISKALY_API_KEY: string, FISKALY_API_SECRET: string, FISKALY_BASE_URL: string;
+
+beforeEach(() => {
+    // @ts-ignore
+    FISKALY_SERVICE_URL = null;
+    FISKALY_API_KEY = 'test_9b92hretadf7nbl9gotci0188_development';
```
Comment (prempador) [inline tests/FiskalyClientTestClientLib.ts]: Please don't have credentials in code.

**Review by prempador — CHANGES_REQUESTED** [review]

File: tests/FiskalyClientTestClientLib.ts (hunk 12 above)
Comment (mihajlon97) [reply tests/FiskalyClientTestClientLib.ts]: Fixed now with .env variables.

**Review by prempador — APPROVED** [review]

**Pull request merged**